app/controllers/admin/users_controller.rb

```diff
@@ -37,16 +37,14 @@ def update
 
     def upload
       begin
-        data = process_file_upload
+        process_file_upload(['.csv'])
       rescue StandardError => e
         flash_message(:error, e.message)
       else
-        if data[:type] == '.csv'
-          @current_job = UploadUsersJob.perform_later(EndUser,
-                                                      params[:upload_file].read,
-                                                      params[:encoding])
-          session[:job_id] = @current_job.job_id
-        end
+        @current_job = UploadUsersJob.perform_later(EndUser,
+                                                    params[:upload_file].read,
```

Comment:
You've correctly identified a potential problem with `.read` calls when given a binary file (or just a file with a different text encoding, but that's a separate matter).

Looking at the implementation of `process_file_upload`, a key difference between `csv` and `yml` handling is that in the latter, there's a `read` call already made, with appropriate error handling if the `read` fails (presumably---you can test this). The former doesn't do this; I think this decision was originally made because it's more memory-efficient to parse a CSV file line-by-line rather than first reading the entire file into memory first (which `.read` does). However, this then makes error handling a bit more challenging, as it becomes spread out throughout the codebase (or just doesn't happen, e.g. here).

So instead, you can make the following change to `process_file_upload`: modify the CSV handling to call `read` on the uploaded file (mimic the YML case), and store the contents in the returned dict as `:contents`, similar again to the YML case. Then modify all places where CSV data is used to instead use `data[:content]` instead. This will also involve more cascading changes across various `from_csv` (and possibly other) methods, but at the end will result in better and more consistent error handling!